libraries/AP_Scripting/applets/quadplane_terrain_avoid.lua

```diff
@@ -0,0 +1,1168 @@
+    local height
+    local search_location = start_location:copy()
+    search_location:change_alt_frame(mavlink.ALT_FRAME.ABSOLUTE)
+
+    local base_height = terrain:height_amsl(search_location, true)
+
+    while search_distance > 0 do
+        search_location:offset_bearing(search_bearing, terrain_spacing)
+        climb = climb + search_ratio * terrain_spacing
+
+        height = terrain:height_amsl(search_location, true)
+        if height ~= nil then
+            local rise = (height - base_height) - climb
+            if rise > highest_rise then
+                local highest_alt_cm = math.floor(math.min(height * 100.0, 320000))
+                highest_rise = rise
+                highest_location = search_location:copy()
+                highest_location:alt(highest_alt_cm)
+            end
+        end
+
+        search_distance = search_distance - terrain_spacing
+    end
+
+    return highest_location
+end
+
+-- returns required pitch to avoid hitting something between here and the next waypoint or other destination such as
+-- home location for RTL
+terrain_approaching = function(clearance)
+    if current_location == nil then
+        gcs:send_text(MAV_SEVERITY.NOTICE, SCRIPT_NAME_SHORT .. string.format(": current_location NIL") )
+        return
+    end
+    local wp_distance = current_location:get_distance(current_location_target)
+    local wp_bearing = math.deg(current_location:get_bearing(current_location_target))
+    local pitch_required = 0
+    local alt_required_amsl = -1
+    local highest_location
+    local highest_alt_difference = 0.0
+    local current_location_amsl = current_location:copy()
+    current_location_amsl:change_alt_frame(mavlink.ALT_FRAME.ABSOLUTE)
+
+    if wp_distance == nil then
+        gcs:send_text(MAV_SEVERITY.NOTICE, SCRIPT_NAME_SHORT .. string.format(": wp_distance NIL") )
+        return 0.0, -1.0, 0.0
+    end
+
+    highest_location = terrain_lookahead(current_location, wp_bearing, wp_distance, 0.5 * tecs_climb_max / groundspeed_current)
+    if highest_location == nil then
+        return 0.0, -1.0, 0.0
+    end
+
+    -- need to know how far ahead the highest location is and how much higher than the current 
+    -- altitude to calculate a minimum required glide slope (which TECS already does)
+    highest_location:change_alt_frame(mavlink.ALT_FRAME.ABSOLUTE)
+    local altitude_difference =  (highest_location:alt() * 0.01) + clearance - (current_location_amsl:alt() * 0.01)
+    if altitude_difference > 0 then
+        -- what is the pitch up requried to acheive that altitude?
+        local highest_distance = current_location_amsl:get_distance(highest_location)
+        pitch_required = math.deg(math.atan(altitude_difference,highest_distance))
+        -- the target location we need to hit needs to be AMSL to ensure we fly above terrain
+        alt_required_amsl = highest_location:alt() * 0.01
+        highest_alt_difference = altitude_difference
+    end
+
```

Comment:
What if you are currently in a dive? You ignore the time it takes to recover from the dive, pitch up, and how much altitude you lose to switch to the max pitch? I guess handling dynamic constraints could be added as a follow up.

Reply:
It would have to be. I currently assume level flight.